packages/@aws-cdk/cdk/lib/core/tag-manager.ts

```diff
@@ -0,0 +1,225 @@
+import { Tag } from '../cloudformation/tag';
+import { Construct } from './construct';
+import { Token } from './tokens';
+
+/**
+ * ITaggable indicates a entity manages tags via the `tags` property
+ */
+export interface ITaggable {
+    tags: TagManager,
+}
+
+/**
+ * Properties Tags is a dictionary of tags as strings
+ */
+export interface Tags {
+    [key: string]: string
+}
+
+/**
+ * A an object of tags with value and properties
+ *
+ * This is used internally but not exported
+ */
+interface FullTags {
+    [key: string]: {value: string, props?: TagProps};
+}
+
+/**
+ * Properties for a tag
+ */
+export interface TagProps {
+    /**
+     * If true all child taggable `Constructs` will receive this tag
+     *
+     * @default true
+     */
+    propagate?: boolean;
+
+    /**
+     * If set propagated tags from parents will overwrite tags in children
+     *
+     * @default false
+     */
+    propagateOverride?: boolean;
+
+    /**
+     * If set this tag will overwrite existing tags
+     *
+     * @default true
+     */
+    overwrite?: boolean;
+}
+
+/**
+ * Properties for removing tags
+ */
+export interface RemoveProps {
+    /**
+     * If true prevent this tag form being set via propagation
+     *
+     * @default true
+     */
+    blockPropagate?: boolean;
+}
+
+/**
+ * TagManager facilitates a common implementation of tagging for Constructs.
+ *
+ * Each construct that wants to support tags should implement the `ITaggable`
+ * interface and properly pass tags to the `Resources` (Cloudformation) elements
+ * the `Construct` creates using `toCloudformation()` for lazy evaluations
+ */
+export class TagManager {
+
+    public static readonly DEFAULT_TAG_PROPS: TagProps = {
+        propagate: true,
+        propagateOverride: false,
+        overwrite: true
+    };
+
+    /**
+     * Checks if the object implements the `ITaggable` interface
+     */
+    public static isTaggable(taggable: ITaggable | any): taggable is ITaggable {
+        return ((taggable as ITaggable).tags !== undefined);
+    }
+
+    /*
+     * Internally tags will have properties set
+     */
+    private readonly _tags: FullTags = {};
+
+    /*
+     * Tags that will be reomved during `tags` method
+     */
+    private readonly blockedTags: string[] = [];
+
+    constructor(private readonly parent: Construct, initialTags: Tags  = {}) {
```

Comment:
`parent` was confusing to me, since this class is not actually a Construct itself, so does not participate in the construct tree.

Can you call it something like `construct` or `taggableConstruct` or something?

Reply:
Ah, I see, Elad suggested this name? I still would prefer something like `associatedConstruct` or something, but I can also live with `parent`.